src_c/surface.c

```diff
@@ -2105,6 +2109,193 @@ surf_blits(pgSurfaceObject *self, PyObject *args, PyObject *keywds)
     return RAISE(PyExc_TypeError, "Unknown error");
 }
 
+#define UBLITS_ERR_TUPLE_REQUIRED 11
+
+static PyObject *
+surf_ublits(pgSurfaceObject *self, PyObject *args, PyObject *keywds)
+{
+    SDL_Surface *src, *dest = pgSurface_AsSurface(self);
+    SDL_Rect *src_rect, temp;
+    temp.x = 0;
+    temp.y = 0;
+    PyObject *srcobject = NULL, *argpos = NULL;
+    int result;
+    SDL_Rect dest_rect;
+    int sx, sy;
+    int the_args = 0;
+
+    PyObject *blitsequence;
+    PyObject *iterator = NULL;
+    PyObject *item = NULL;
+    PyObject *special_flags = NULL;
+    PyObject *ret = NULL;
+    PyObject *retrect = NULL;
+    Py_ssize_t itemlength;
+    int doreturn = 1;
+    int bliterrornum = 0;
+    static char *kwids[] = {"blit_sequence", "special_flags", "doreturn",
+                            NULL};
+    if (!PyArg_ParseTupleAndKeywords(args, keywds, "O|Oi", kwids,
+                                     &blitsequence, &special_flags, &doreturn))
+        return NULL;
+
+    if (doreturn) {
+        ret = PyList_New(0);
+        if (!ret)
+            return NULL;
+    }
+
+    if (!PyIter_Check(blitsequence) && !PySequence_Check(blitsequence)) {
+        bliterrornum = BLITS_ERR_SEQUENCE_REQUIRED;
+        goto bliterror;
+    }
+
+    if (special_flags) {
+        if (!pg_IntFromObj(special_flags, &the_args)) {
+            bliterrornum = BLITS_ERR_MUST_ASSIGN_NUMERIC;
+            goto bliterror;
+        }
+    }
+
+    iterator = PyObject_GetIter(blitsequence);
```

Comment:
An idea. We are accepting everything that is an iterable. I've seen that you can create a `PySequence_Fast`: https://docs.python.org/3/c-api/sequence.html#c.PySequence_Fast
Couldn't we convert the blit_sequence to `PySequence_Fast` (or maybe to a PySequence_Fast_ITEMS) and use PySequence_Fast functions like `PySequence_Fast_GET_ITEM` to get items from the sequence? Would't that be faster? We'd need a `Py_ssize_t counter` and then we'd have a normal for(counter=0; counter < sequence_length; counter++) to access the items?

Reply:
You'd have to assume just a list or a tuple is passed as `blit_sequence` function parameter, and so you'd restrict the function's purpose even more, but you'd get (possibly?) even more performance because at that point you'd just be looping over an array ( if you use `PySequence_Fast_ITEMS` ). This function could very well become a very optimized version of blits() if you're willing to restrict the function's behaviour, but would probably take some time in that case.

Reply:
But then you lose the possible speed benefit of using a generator. Since currently you can pass in "a list" without ever having to construct it. You could have it both ways by trying to use PySequence_Fast and then using an iterator strategy? Not sure what would be worth it with performance.

Also if you're not in the pygame community discord you should join, we have a "contributing" channel for pygame development. https://discord.com/invite/ZuB2RySPRJ

Reply:
> 
Yes i'll ty and join the discord.
Yeah something more along the "both ways" would be ideal. I could try it and post some tests... 
and i've ran a quck test and there seems to be no real difference whether you construct a list on the fly or use a generator, there may still be a difference if you don't construct the list at all just like you said:
WITH IMPLICIT GENERATOR
blits: 0.014914000001226668
ublits: 0.012218600000778679

WITH EXPLICIT GENERATOR
blits: 0.012284700000236626
ublits: 0.012045800000123563

WITH LIST:
blits: 0.012100500000087777
ublits: 0.011486600000353064

Code i used:
```python
print("WITH IMPLICIT GENERATOR")
print("blits: ", end="")
print(timeit(f"win.blits((IMAGE, POSITIONS[i]) for i in range(NUMSURFS))", number=100, globals=GLOB))
print("ublits: ", end="")
print(timeit(f"win.ublits((IMAGE, POSITIONS[i]) for i in range(NUMSURFS))", number=100, globals=GLOB))

print("\nWITH EXPLICIT GENERATOR")
print("blits: ", end="")
print(timeit(f"win.blits(((IMAGE, POSITIONS[i]) for i in range(NUMSURFS)))", number=100, globals=GLOB))
print("ublits: ", end="")
print(timeit(f"win.ublits(((IMAGE, POSITIONS[i]) for i in range(NUMSURFS)))", number=100, globals=GLOB))

print("\nWITH LIST:")
print("blits: ", end="")
print(timeit(f"win.blits([(IMAGE, POSITIONS[i]) for i in range(NUMSURFS)])", number=100, globals=GLOB))
print("ublits: ", end="")
print(timeit(f"win.ublits([(IMAGE, POSITIONS[i]) for i in range(NUMSURFS)])", number=100, globals=GLOB))
```

Reply:
I've done more testing and came to the conclusion that using lists is overall faster than using generators, and that makes sense because when you call `PyIter_Next(iterator)`, you get the next element from the sequence, and in the case of generators there's extra operations to get the next value that can be arbitrarily complex. 

That being said I'd like to iterate over this idea of using PySequence_Fast and discuss why we should move to a simple C array of PyObject*. But first i'd like to respond to what you said:

> ...You could have it both ways by trying to use PySequence_Fast and then using an iterator strategy? Not sure what 
 would be worth it with performance.

- You can't have it both ways because `PySequence_Fast` objects can only be created converting a List or Tuple object. Therefore you cannot convert a generator object.
-  `PySequence_Fast` objects are intended for accessing the sequence's data fields directly, with either a PySequence_Fast_GET_ITEM or a PySequence_ITEM, so an iterator strategy could be overkill in this case.

Let me unpack what we have to do right now  vs. what we could have into two lists here

RIGHT NOW:
1. Have to declare, assign, manage and DECREF a `PyObject *iterator`
2. Check the validity of the blit sequence with `PyIter_Check` and `PySequence_Check`
4. call `PyIter_Next(iterator)` on every loop
5. `Py_DECREF(item);` on every loop because `PyIter_Next` gives a new reference
6. In the end we `Py_DECREF(iterator);`

INSTEAD WE COULD HAVE:
1. Have a `blit_sequence = PySequence_Fast(args[0], errcode)`
2. Convert the `blit_sequence ` into an array of the underlying sequence objects with `PySequence_Fast_ITEMS(blit_sequence)`
3. Loop over the array with `for (seq_counter = 0; seq_counter < sequencelength; seq_counter++)`, so no iterators
4. Have `item = f_blitsequence[seq_counter]` and just use it with no DECREF

## Positives
- We should have less code to manage and an easier to understand function
- We should get a modest performance uplift as we are no longer using iterators, but looping through an array
- If not we have a less noisy (meaning little negative performance spikes), and more reliable function

## Negatives
- Having to use lists or tuples

In my opinion, given that this function already simplifies the `blit_sequence `by only having pairs (Surface, dest), it only makes sense to have only lists or tuples, because even in the worst case scenario the user would have to write something like 
`surf.ublits([(img, pos) for pos in positions], 0, 1)`
instead of 
`surf.ublits(((img, pos) for pos in positions), 0, 1)`
and the same is true if a uses uses a function to return a generator. Simple generators are easy to substitute with list comprehensions and complex ones don't really fit this use case.
So i don't think that swapping two "( )" for a "[ ]" is really all that bad.

Reply:
You can have it both ways, you can have a code path for one and a code path for the other.

Generators "overhead" constructing the next element is irrelevant because the list would have to do that work as well.

> So i don't think that swapping two "( )" for a "[ ]" is really all that bad.

A small change in syntax, but a large change in what's actually happening.

Reply:
Although a quick personal test doesn't seem very favorable on generators. It beat the list comp, but not by a significant margin.

Reply:
can i get the code snippet so i can replicate this?

Reply:
> You can have it both ways, you can have a code path for one and a code path for the other.

code would start to get very long i'm just saying, because on one you'd have a while(item=PyIter_Next(blit_sequence))... and on the other a for loop over an array.

Reply:
> Although a quick personal test doesn't seem very favorable on generators. It beat the list comp, but not by a significant margin.

Did you use any blend flags over the default one?

Reply:
or are you just talking about difference between gen and list relative perf